tlatools/org.lamport.tlatools/test/tlc2/util/BufferedRandomAccessFileFuzzTest.java

```diff
@@ -0,0 +1,553 @@
+        int stride = list.size() / 2;
+
+        while (stride > 0) {
+            boolean listModified = false;
+
+            for (int i = 0; i < list.size(); i += stride) {
+                int cutStart = i;
+                int cutEnd = Math.min(i + stride, list.size());
+                List<T> candidate = append(
+                        list.subList(0, cutStart),
+                        list.subList(cutEnd, list.size()));
+//                System.out.println("Attempting to remove [" + cutStart + ", " + cutEnd + ") = " + list.subList(cutStart, cutEnd) + " ==> " + candidate + "...");
+                if (test.test(candidate)) {
+//                    System.out.println(" --> REMOVED!");
+                    list = candidate;
+                    i -= stride;
+                    listModified = true;
+                }
+            }
+
+            if (!listModified) {
+                stride /= 2;
+            }
+        }
+
+        return list;
+    }
+
+    private static <T> List<T> append(List<T> a, List<T> b) {
+        if (a.isEmpty()) {
+            return b;
+        }
+        if (b.isEmpty()) {
+            return a;
+        }
+        List<T> result = new ArrayList<>(a.size() + b.size());
+        result.addAll(a);
+        result.addAll(b);
+        return result;
+    }
+
+    private interface IOPredicate<T> {
+        boolean test(T value) throws IOException;
+    }
+
+    private static class RunResult {
+        final java.util.List<FileOperation<?>> ops;
+        final int successes;
+        final Object expected;
+        final Object actual;
+
+        public RunResult(java.util.List<FileOperation<?>> ops, int successes, Object expected, Object actual) {
+            this.ops = ops;
+            this.successes = successes;
+            this.expected = expected;
+            this.actual = actual;
+        }
+
+        public boolean ok() {
+            return Objects.equals(expected, actual);
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            RunResult runResult = (RunResult) o;
+            return successes == runResult.successes &&
+                    Objects.equals(expected, runResult.expected) &&
+                    Objects.equals(actual, runResult.actual);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(successes, expected, actual);
+        }
+
+        @Override
+        public String toString() {
+            return "RunResult{" +
+                    "opsExecuted=" + successes +
+                    ", expected=" + expected +
+                    ", actual=" + actual +
+                    '}';
+        }
+    }
+
+    private RunResult checkOps(java.util.List<FileOperation<?>> ops) throws IOException {
```

Comment:
I might recommend calling this `checkLegalOps`.  I think `check` and `legal` make as much sense as any names, but I did have a moment of confusion between the purpose of the two.  This name might make clear that the function's domain is only legal operations, and we should therefore always expect that we're going to have to check their legality first.

Reply:
Now that you mention it, I think "legal" should be "well-defined". I used the two terms interchangeably, but I think "well-defined" is a better name for it.

Reply:
Nice, yeah, I think this is a much clearer change!